pmd-core/src/main/java/net/sourceforge/pmd/cache/AbstractAnalysisCache.java

```diff
@@ -114,6 +120,23 @@ public void checkValidity(final RuleSets ruleSets, final ClassLoader classLoader
         ruleMapper.initialize(ruleSets);
     }
 
+    private long computeClassLoaderHash(final URLClassLoader classLoader) {
+        final Adler32 adler32 = new Adler32();
+        for (final URL url : classLoader.getURLs()) {
+            try (CheckedInputStream inputStream = new CheckedInputStream(url.openStream(), adler32)) {
+                // Just read it, the CheckedInputStream will update the checksum on it's own
+                while (IOUtils.skip(inputStream, Long.MAX_VALUE) == Long.MAX_VALUE) {
+                    // just loop
+                }
+            } catch (final IOException e) {
+                // Can this even happen?
```

Comment:
It could happen, but usually not so often: Suppose you run PMD in parallel with building some dependencies that are on the auxclasspath. It might happen then, that the file, that we have currently opened, it replaced on the fly. Under linux, we might keep the file handle, but the content is changing - if the file size shrinks, we could maybe try to read beyond the end of the file...

I've noticed a similar behavior when building PMD snapshot (mvn clean install) and using the very same snapshot version as a dependency to the maven-pmd-plugin: The original pmd dependencies are loaded from "\~/.m2/repository/..." when m-pmd-p is analyzing pmd-core. But after that, pmd-core is installed and replaces the file in "\~/.m2/repository/...". Depending on the change to the file, I've experienced spurious ClassNotFound exceptions, since java got an error reading the jar file...

Reply:
Thanks for the input, good to know!